// @rmwc/typography/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Typography = undefined;

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _base = require('@rmwc/base');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
// eslint-disable-next-line no-unused-vars


var TypographyRoot = (0, _base.simpleTag)({
  displayName: 'Typography',
  defaultProps: {
    use: undefined
  },
  tag: 'span',
  classNames: function classNames(props) {
    return [_defineProperty({}, 'mdc-typography--' + props.use, props.use)];
  },
  consumeProps: ['use']
});

/**
 * The Typography Component
 */
var Typography = exports.Typography = function Typography(props) {
  return React.createElement(TypographyRoot, props);
};

Typography.displayName = 'Typography';

exports.default = Typography;
