// @rmwc/base/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _simpleTag = require('./simpleTag');

Object.defineProperty(exports, 'simpleTag', {
  enumerable: true,
  get: function get() {
    return _simpleTag.simpleTag;
  }
});

var _withTheme = require('./withTheme');

Object.defineProperty(exports, 'withTheme', {
  enumerable: true,
  get: function get() {
    return _withTheme.withTheme;
  }
});

var _noop = require('./utils/noop');

Object.defineProperty(exports, 'noop', {
  enumerable: true,
  get: function get() {
    return _noop.noop;
  }
});

var _randomId = require('./utils/randomId');

Object.defineProperty(exports, 'randomId', {
  enumerable: true,
  get: function get() {
    return _randomId.randomId;
  }
});

var _withFoundation = require('./withFoundation');

Object.defineProperty(exports, 'withFoundation', {
  enumerable: true,
  get: function get() {
    return _withFoundation.withFoundation;
  }
});
Object.defineProperty(exports, 'syncFoundationProp', {
  enumerable: true,
  get: function get() {
    return _withFoundation.syncFoundationProp;
  }
});

var _foundationComponent = require('./foundation-component');

Object.defineProperty(exports, 'FoundationComponent', {
  enumerable: true,
  get: function get() {
    return _foundationComponent.FoundationComponent;
  }
});

var _component = require('./component');

Object.defineProperty(exports, 'Component', {
  enumerable: true,
  get: function get() {
    return _component.Component;
  }
});
